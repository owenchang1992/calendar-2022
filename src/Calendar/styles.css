.calendar-container {
  width: 350px;
  display: flex;
  flex-direction: column;
}

.calendar-header {
  height: 44px;
  margin-bottom: 16px;
  display: grid;
  grid-template-columns: 44px auto 44px;
}

.v-h-center {
  display: flex;
  justify-content: center;
  align-items: center;
}

.icons {
  background-color: var(--color-white);
}

.icons:hover {
  background-color: var(--color-light-gray);
}

.calendar-dates {
  display: grid;
  grid-template-columns: repeat(7, 1fr);
  grid-template-rows: repeat(6, 1fr);
  justify-content: center;
  align-items: center;
}

.today {
  background-color: var(--color-yellow);
}

.active {
  background-color: var(--color-blue);
}

.date {
  height: 36px;
}

.date:hover {
  background-color: var(--color-light-gray);
}

.non-current-month {
  color: var(--color-gray);
  cursor: not-allowed
}

