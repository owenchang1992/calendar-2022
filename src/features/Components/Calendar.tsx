import { BiChevronLeft, BiChevronRight } from 'react-icons/bi'
import './Calendar.css'
import useCalendar from '../hooks/useCalendar'

const Calendar = () => {
  const [
    calendar,
    monthForward,
    monthBackward,
    selectDate,
  ] = useCalendar();

  const getBackgroundColor = (date: { title: number; isNonCurrentMonth: boolean; isToday: boolean; isActive: boolean; }) => {
    if (date.isNonCurrentMonth) return 'non-current-month' 
    if (date.isActive) return 'active' 
    if (date.isToday) return 'today'
    return ''
  }

  return (
    <div className='calendar-container'>
      <div className="calendar-header">
        <div className='v-h-center icons' onClick={monthBackward}>
          <BiChevronLeft />
        </div>  
        <div className='calendar-month v-h-center'>
          {`${calendar.year}年${calendar.month + 1}月`}
        </div>
        <div className='v-h-center icons' onClick={monthForward}>
          <BiChevronRight />
        </div>
      </div>
      <div className="calendar-dates">
        {
          calendar.dates.map((date) => (
            <div
              className={`date v-h-center ${getBackgroundColor(date)}`}
              onClick={() => selectDate(date)}
              key={`${date.isNonCurrentMonth ? calendar.month : 'isNonCurrentMonth'}-${date.title}`}
            >
              {`${date.title}日`}
            </div>
          ))
        }
      </div>
    </div>
  )
}

export default Calendar

