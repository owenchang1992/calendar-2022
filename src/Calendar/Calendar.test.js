import Calendar from './index'
import renderer from 'react-test-renderer';

describe('Test Calendar', () => {
  const RealDate = Date

  Date = class extends RealDate {
    constructor(...args) {
      if (args.length === 0) return new RealDate(2022, 7, 22)
      return new RealDate(...args)
    }
  }

  afterEach(() => {
    Date = RealDate
  })

  it('render 2022/8/22', () => {
    const tree = renderer
      .create(<Calendar />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  })
});
