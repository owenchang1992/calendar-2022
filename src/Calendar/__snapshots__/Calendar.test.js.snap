// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Calendar render 2022/8/22 1`] = `
<div
  className="calendar-container"
>
  <div
    className="calendar-header"
  >
    <div
      className="v-h-center icons"
      onClick={[Function]}
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13.293 6.293 7.586 12l5.707 5.707 1.414-1.414L10.414 12l4.293-4.293z"
        />
      </svg>
    </div>
    <div
      className="calendar-month v-h-center"
    >
      2022年8月
    </div>
    <div
      className="v-h-center icons"
      onClick={[Function]}
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10.707 17.707 16.414 12l-5.707-5.707-1.414 1.414L13.586 12l-4.293 4.293z"
        />
      </svg>
    </div>
  </div>
  <div
    className="calendar-dates"
  >
    <div
      className="date v-h-center non-current-month"
      onClick={[Function]}
    >
      31日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      1日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      2日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      3日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      4日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      5日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      6日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      7日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      8日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      9日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      10日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      11日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      12日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      13日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      14日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      15日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      16日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      17日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      18日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      19日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      20日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      21日
    </div>
    <div
      className="date v-h-center today"
      onClick={[Function]}
    >
      22日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      23日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      24日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      25日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      26日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      27日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      28日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      29日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      30日
    </div>
    <div
      className="date v-h-center "
      onClick={[Function]}
    >
      31日
    </div>
    <div
      className="date v-h-center non-current-month"
      onClick={[Function]}
    >
      1日
    </div>
    <div
      className="date v-h-center non-current-month"
      onClick={[Function]}
    >
      2日
    </div>
    <div
      className="date v-h-center non-current-month"
      onClick={[Function]}
    >
      3日
    </div>
  </div>
</div>
`;
